### PR TITLE
issues #1907 and #1964 - strip the query from originalRequestUri before parsing and fix baseUrl for whole-system queries

### PR DESCRIPTION
**Pull request by lmsurpre**
1. Previously I updated our UriBuilder for this case (and introduced tests
to ensure it is properly functioning). However, I missed the fact that
FHIRRestServletFilter parses the originalRequestUri header value into a
URI and, when this fails, it simply reverts to the actual URL of the
request. Therefor, when the request contained a character like `|`, this
header value was never being used in downstream processing like
UriBuilder.buildSelfUri.toSearchSelfUri().

2. Set originalRequestUriHeaderName to "X-FHIR-FORWARDED-URL" in our
default fhir-server-config (so that this can get tested by our default
CI without flags)
3. Added OriginalRequestRewriteServerTest to ensure this feature is
working as expected
4. Expanded SearchAllTest to include a check for the fullUrl entries in the
response bundle
5. Fix FHIRRestHelper.getRequestBaseUri(type) in two cases:
  * for certain cases we were setting the baseUri to requestUri (losing our strip of the query string)
  * strip the `/_search` from the end of the whole-system search URL

6. Also performed some minor cleanup.

Signed-off-by: Lee Surprenant <lmsurpre@us.ibm.com>

### CODE REVIEW

**Review by michaelwschroeder — APPROVED** [review]
LGTM

File: fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java [hunk 1]
```diff
@@ -316,41 +316,51 @@ private String getRequestURL(HttpServletRequest request) {
     /**
      * Get the original request URL from either the HttpServletRequest or a configured Header (in case of re-writing proxies).
      *
+     * If configured as a header, the header value will only be used for the protocol, server name, port number, and server path;
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/filter/rest/FHIRRestServletFilter.java:319]: Copyright year

**Review by prb112 — APPROVED** [review]
LGTM - one file needs a copyright year update

File: fhir-server/liberty-config/config/default/fhir-server-config.json [hunk 2]
```diff
@@ -4,6 +4,7 @@
         "core": {
             "tenantIdHeaderName": "X-FHIR-TENANT-ID",
             "datastoreIdHeaderName": "X-FHIR-DSID",
+            "originalRequestUriHeaderName": "X-FHIR-FORWARDED-URL",
```
Comment (prb112) [inline fhir-server/liberty-config/config/default/fhir-server-config.json:7]: one additional thought here... we prefix with X-FHIR, it's generally discouraged. I do like the consistency with the other headers, and wanted to just leave this comment for posterity/ future thought.

https://tools.ietf.org/html/rfc6648
https://stackoverflow.com/a/3561399/1873438

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java [hunk 3]
```diff
@@ -2841,14 +2841,16 @@ private String getRequestBaseUri(String type) throws Exception {
...
+        }
+
+        // Strip any path segments for whole-system interactions (in case of whole-system search, "Resource" is passed as the type)
+        if (type == null || type.isEmpty() || "Resource".equals(type)) {
+            if (baseUri.endsWith("/_search")) {
                 baseUri = baseUri.substring(0, baseUri.length() - "/_search".length());
+            } else if (baseUri.endsWith("/_history")) {
+                baseUri = baseUri.substring(0, baseUri.length() - "/_history".length());
             } else if (baseUri.contains("/$")) {
                 baseUri = baseUri.substring(0, baseUri.lastIndexOf("/$"));
```
Comment (prb112) [inline fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2855]: Thiss one gets called when processing a bundle.

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/OriginalRequestRewriteServerTest.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
+/*
+ * (C) Copyright IBM Corp. 2017, 2020
+ *
```
Comment (prb112) [inline fhir-server-test/src/test/java/com/ibm/fhir/server/test/OriginalRequestRewriteServerTest.java:3]: 2021

**Review by prb112 — APPROVED** [review]
LGTM (minus CI failures)

**Comment by prb112** [comment]
```
Error:  Failures: 
Error:    OriginalRequestRewriteServerTest.testCreatePatient_LocationRewrite:53 https://localhost:9443/fhir-server/api/v4/Patient/177b25f6c63-fdb38467-b6c6-4950-851a-098907ae3d8d/_history/1 should begin with https://frontend-url/base expected [true] but found [false]
```

File: fhir-server/liberty-config/config/default/fhir-server-config.json (hunk 2 above)
Comment (lmsurpre) [reply fhir-server/liberty-config/config/default/fhir-server-config.json:7]: Interesting.  I tend to agree with the first commenter on that answer:
> Just as there are many kids that will never end up as professional athletes, many custom headers will never end up as standards. I'm inclined to keep the "X-" on those.

Especially given that we already have the two others with the "X-FHIR" prefix.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java (hunk 3 above)
Comment (lmsurpre) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2855]: Not sure if you were suggesting that something should change or not; I mostly just repositioned this and made it run when the type is "Resource".  Without this change, I found that the fullUrl in our search response bundles contained a trailing `_search` (when the search was invoked via `POST [base]/_search`).

File: fhir-server-test/src/test/java/com/ibm/fhir/server/test/OriginalRequestRewriteServerTest.java (hunk 4 above)
Comment (lmsurpre) [reply fhir-server-test/src/test/java/com/ibm/fhir/server/test/OriginalRequestRewriteServerTest.java]: fixed

File: fhir-server/liberty-config/config/default/fhir-server-config.json (hunk 2 above)
Comment (prb112) [reply fhir-server/liberty-config/config/default/fhir-server-config.json:7]: I like the consistency, but wanted to bring it up.

File: fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java (hunk 3 above)
Comment (prb112) [reply fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java:2855]: Got it.

**Review by michaelwschroeder — APPROVED** [review]
LGTM

**Comment by d0roppe** [comment]
LGTM - when passing in the full URI

**Review by d0roppe — APPROVED** [review]
LGTM - when passing in the full URI

**Pull request merged**